### DISCUSSION TITLE
Why is the loss function of fastflow negative, What are the meanings of pixel_F1Score  and pixel_AUROC？Thanks

### DISCUSSION THREAD

**Discussion by leibaoer** [open-edge-platform/anomalib#579] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/113488512/190044998-994ee405-d4fc-4180-9eff-e63756b1bc2b.png)

**Comment by leibaoer** [comment]
hello

**Comment by cytotoxicity8** [comment, marked as answer]
The loss is likelihood as general normalizing flow models. It will be helpful to search concepts of normalizing flow. In addition, fastflow predicts whether the pixel is anomaly. Pixel AUROC means the localization(segmentation) perfomance.

**Comment by ashwinvaidya17** [comment]
@leibaoer sorry for the late response. As @cytotoxicity8 mentioned, fastflow computes the loss by maximizing the negative log jacobian. You can see it defined here https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/models/fastflow/loss.py#L27. Section 3.3 of the fastflow paper (https://arxiv.org/pdf/2111.07677.pdf) might make it more clear.

**Reply by leibaoer** [reply]
Thanks,I am a student comes from China,so my English is not very good.I want to ask you another question, and look forward to your reply.My defects can be displayed on the heatmaps, but they can't be circled on the pre mask. What parameters do I need to adjust? My test results are not very good,like these.
![image](https://user-images.githubusercontent.com/113488512/192074267-c8797724-8da4-48de-bbb9-5a9e5ac81291.png)
![image](https://user-images.githubusercontent.com/113488512/192074269-a829561c-faaf-4229-a9bd-58156f541a1b.png)
![image](https://user-images.githubusercontent.com/113488512/192074271-f0f22c73-5a82-465f-8b90-95b3a05a53cd.png)
![image](https://user-images.githubusercontent.com/113488512/192074272-f65b844c-1a23-4614-ac59-96726974bdcc.png)

**Reply by ashwinvaidya17** [reply]
The segmentation maps are generated from the threshold. So while the heatmap might show something the threshold determines the value above which the predicted values are considered as anomalies. Normalization ensures that this threshold is centred to 0.5. So, when generating the segmentation masks the anomalous regions are the ones with values greater than 0.5. 
1. Since this threshold is calculated from the validation images, I would suggest  adding more images with ground truth masks in your dataset.
2. You can also have a look at the adaptive threshold class https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/utils/metrics/adaptive_threshold.py#L10. Currently, the threshold is computed based on optimal f1 score. Maybe you use a different metric here.
